### DISCUSSION TITLE
Run bmat in solve

### DISCUSSION THREAD

**Discussion by kinnala** [kinnala/scikit-fem#736] [Ideas]
I think `solve` could automatically apply `scipy.sparse.bmat` on lists to allow syntax such as
```python
solve([[A, B.T], [B, None]], [f, None])
```
Custom function is needed for RHS to have similar syntax for numpy arrays but it should be easy enough?

**Comment by kinnala** [comment]
What about `codense` et al? Should that do it also?

**Reply by gdmcbain** [reply]
I suppose in `condense`, it's usually only the first row that needs condensing in these saddle-point problems; e.g. for Stokes that's the velocity and one doesn't apply Dirichlet conditions to the pressure.

Is the idea here syntactic sugar or is there some way to exploit the block structure; e.g. for a saddle-point like your example?

**Reply by kinnala** [reply]
I was thinking syntactic sugar, however, I'm not sure how condensing etc. would work. Should `D` and `I` also be a list?

**Reply by kinnala** [reply]
Such as
```python
solve(*condense([[A, B.T], [B, None]], [f, None], I=[abasis.get_dofs(), None]))
```

**Reply by gdmcbain** [reply]
Hmm.  My first thought was to accept either a `list[ndarray]` or just an `ndarray` and in the latter case to interpret the indices in the big matrix block-assembled by `bmat`, but then I realized that this might be awkward if the indices are passed as an &lsquo;array-like&rsquo; list.  Perhaps it would suffice to check that the `len` of `I` or `D` matched that of `A` or `A[0]`, in which case it should be interpreted in the first way, i.e. block-wise, and otherwise try the second way?

But then maybe being a bit stricter and insisting on the form you've shown (`I=[abasis.get_dofs(), None]`) is reasonable.  It's not too verbose and it is clearer and less ambiguous.  On reflection, I think the latter might be better than accepting a `Union` type.

I suppose that this would be done in

https://github.com/kinnala/scikit-fem/blob/bc57d968e56e6b89a99e35eac26ef7bc81b7a46a/skfem/utils.py#L248
